fiftyone/utils/labelstudio.py

```diff
@@ -0,0 +1,835 @@
+
+Run Label Studio:
+    1. As a python package:
+        pip install label-studio
+        label-studio start
+    2. As a docker image:
+        docker run -d -p 8080:8080 \
+            --name labelstudio \
+            -v /tmp/labelstudio:/data \
+            labelstudio/labelstudio
+
+Set environment variables:
+    export FIFTYONE_LABELSTUDIO_URL=<url>
+    export FIFTYONE_LABELSTUDIO_TOKEN=<token>
+
+Annotate:
+    dataset = fo.load_dataset("quickstart")
+    dataset.annotate(<anno_key>, label_field=<field_name>,
+                    label_type=<label_type>,
+                    classes=[list, of, classes],
+                    backend="labelstudio")
+    # Note: a new project will be created with a labelling config
+    # that matches label type and classes.
+    # Note: it might take a while because files have to be
+    # uploaded to the Label Studio server.
+
+Load annotations:
+    dataset.load_annotations(<anno_key>)
+    # Note: this will only load the latest annotation per task
+    # if multiple annotations exist for a task.
+"""
+import itertools
+import json
+import logging
+import os
+import random
+import string
+import webbrowser
+from copy import deepcopy
+from datetime import datetime as dt
+from typing import List
+
+import numpy as np
+
+import fiftyone as fo
+import fiftyone.core.utils as fou
+import fiftyone.core.labels as fol
+import fiftyone.utils.annotations as foua
+
+ls = fou.lazy_import(
+    "label_studio_sdk", callback=lambda: fou.ensure_import("label_studio_sdk")
+)
+
+logger = logging.getLogger(__name__)
+
+
+class LabelStudioBackendConfig(foua.AnnotationBackendConfig):
+    """Base class for Label Studio backend configuration.
+
+    Args:
+        name: the name of the backend
+        label_schema: a dictionary containing the description of label fields,
+            classes and attribute to annotate
+        media_field ("filepath"): string field name containing the paths to
+            media files on disk to upload
+        url: the URL of the Label Studio server
+        api_key: the API key to use for authentication
+        project_name: the name of the project to use on the Label Studio server
+
+    """
+
+    def __init__(
+        self,
+        name: str,
+        label_schema: str,
+        media_field="filepath",
+        url="http://localhost:8080",
+        api_key=None,
+        project_name=None,
+        **kwargs,
+    ):
+        super().__init__(name, label_schema, media_field=media_field, **kwargs)
+
+        self.url = url
+        self.project_name = project_name
+
+        self._api_key = (
+            api_key if api_key else os.getenv("FIFTYONE_LABELSTUDIO_TOKEN")
+        )
+        if self._api_key is None:
+            raise ValueError("No Label Studio API key provided")
+
+    @property
+    def api_key(self):
+        return self._api_key
+
+    @api_key.setter
+    def api_key(self, value):
+        self._api_key = value
+
+
+class LabelStudioBackend(foua.AnnotationBackend):
```

Comment:
This class is required to implement the [`recommend_attr_tool()`](https://github.com/voxel51/fiftyone/blob/8cab046e7fe98ad04e37bab70183e03e128904bf/fiftyone/utils/annotations.py#L1834) method which basically defines the type of annotation widget that will be used for a given attribute, generally based on the type of attribute that is being parsed.

Reply:
I haven't added attribute handling yet. Is it strictly necessary at the start?

Reply:
No problem! It makes sense to get everything working first before tackling attributes.

Reply:
are we closing this one for now?

Reply:
So at the moment, if attributes are provided, then `recommend_attr_tool()` will be called. This can happen by default if passing in an existing label field where attributes are parsed automatically. The way to avoid it, you can set `attributes=False`, like this:
```python
dataset.annotate(anno_key, label_field="existing_field", attributes=False, backend="labelstudio")
```

From what I can tell, there's two ways to go about this:
1) Raise a `NotImplementedError` in `LabelStudioBackend.recommend_attr_tool()` that instructs the user to set `attributes=False` until attributes are supported by this backend.
2) Add a flag that automatically ignores all attributes for a given backend [like this](https://github.com/voxel51/fiftyone/commit/fa55781dda7b25dff0432581f3740db43ce18fd1).

While it would be smoother to automatically ignore attributes, I'm a bit worried about that behavior being unexpected for users. What are your thoughts @Rusteam and @brimoor ?

Another alternative is to just add attribute support, but this issue can come up with future backends too so it would be useful to discuss.

Reply:
Raising a `NotImplementedError` is a reasonable thing to do if we plan to implement it soon, but I guess the crux of the decision is that it would be annoying that the simplest syntax:

```py
dataset.annotate(anno_key, label_field="existing_field", backend="labelstudio")
```

would fail since the public `annotate()` API has already decided that `attributes=True` is always the default. Seems like good motivation to get attributes support done too 🤗 

I would prefer to only introduce `supports_attributes` if we think there will be annotation backends that don't support label attributes at all, not just because we haven't implemented it yet.

If we do introduce `supports_attributes`, then we could update the default to `attributes=None` in the public `annotate(...)` API and then [document](https://voxel51.com/docs/fiftyone/user_guide/annotation.html#requesting-annotations) that the default value for the `attributes` parameter will be set to the value of `supports_attributes` property of the annotation backend being used. That would help address Eric's concern of certain backends silently ignoring attributes.

Reply:
Seems like implementing attributes is the only viable solution :)

Reply:
Do I understand correctly that attributes correspond to label tags? Is it usually applied to a label with multiple instances (detections, polygons, etc.) and can be something like `[occluded, partial, etc.]`?
If so, I don't think Label Studio supports per-label annotations. Their [control tags](https://labelstud.io/tags/relation.html) require a label to be attached to input data `toName`. I will double-check with their community

Reply:
Hmm, it seems you're right. From looking into Labelstudio more, it doesn't seem like they have a way to add attributes directly.

In FiftyOne, the idea of an attribute is any field added to a `Label` object other than the `Label.label` representing the class/category. The closest I found are the `Relations` that you linked or a [`TextArea`](https://labelstud.io/tags/textarea.html) (which is really meant for captioning and doesn't support numerical/boolean types that a lot of attributes use) neither of which are a clean way to support attributes. 

I'll dig a bit deeper, but it seems that Labelstudio may actually require `supports_attributes=False` until [this is resolved](https://github.com/heartexlabs/label-studio/issues/794).

Reply:
iWould it beeter to not throw an error by default but rather show a warning that attributes are not supported?

Reply:
@Rusteam Thanks for all the updates you've been making to this! I've opened a PR (https://github.com/Rusteam/fiftyone/pull/1) on your fork into this branch adding the `supports_attributes` updates (as well as a few other things like Labelstudio docs). Take a look at that when you get a chance and let me know if you want to discuss anything there, if not, you can go ahead and merge it in.